---
title: "SeSAMe User Guide"
shorttitle: "sesame guide"
author: "Wanding Zhou"
package: sesame
output: rmarkdown::html_vignette
fig_width: 6
fig_height: 4
vignette: >
  %\VignetteIndexEntry{SeSAMe User Guide}
  %\VignetteEngine{knitr::rmarkdown}
  \usepage[utf8]{inputenc}
---

# Introduction

SeSAMe is an R package for processing Infinium DNA methylation data. SeSAMe currently supports EPIC, HM450 and HM27 platforms.

## To read IDATs from directory
```R
library(sesame)
ssets <- readIDATsFromDir('IDATs/')
```
        
A simple list of "SignalSet"s are returned.
        
Other options of data import include

+ readIDATs - from path
+ readIDATsFromSheet - from data.frame "barcode" column
        
### Background subtraction
    
noob:
```R
library(sesame)
sset <- makeExampleTinyEPICDataSet()
sset.nb <- noob(sset)
```
      
### Dye bias correction

linear scaling
```{r}
library(sesame)
sset <- makeExampleTinyEPICDataSet()
sset.db <- dyeBiasCorr(sset)
```
      
quantile interpolation with Type-I probes
```{r}
sset <- makeExampleTinyEPICDataSet()
sset.db <- dyeBiasCorrTypeINorm(sset)
sset.db
```
      
### Get betas

```{r}
sset <- makeExampleTinyEPICDataSet()
betas <- getBetas(sset)
head(betas)
```
use option quality.mask = TRUE/FALSE to mask probes with mapping issues, SNPs 	and non-uniqueness, described in Zhou 2016 NAR.
use option nondetection.mask = TRUE/FALSE to mask nondetection calls.

Both masks are recommended to ensure data quality and defaulted to TRUE.
        
### Sample/experiment QC

#### sex
```R
inferSex(sset)
```

#### ethnicity
```R
inferEthnicity(sset)
```

#### age
```{r}
betas <- SeSAMeGetExample('HM450.betas.TCGA-2L-AAQA-01A-21D-A38H-05')
predictAgeHorvath353(betas)
```

#### mean intensity
```{r}
sset <- SeSAMeGetExample('EPIC.sset.LNCaP.Rep1')
meanIntensity(sset)
```

#### bisulfite conversion control using [GCT score](https://academic.oup.com/nar/article/45/4/e22/2290930)
```{r}

bisConversionControl(sset)
```

### visualize probes

```{r}
betas <- SeSAMeGetExample('HM450.betas.10TCGAnormalPAAD')
```

visualize probes from a gene
```{r}
visualizeGene('DNMT1', betas, platform='HM450')
```

visualize probes from arbitrary region
```R
visualizeRegion('chr19',10260000,10380000, betas, platform='HM450')
```

visualize by probe names
```R
visualizeProbes(c("cg02382400", "cg03738669"), betas, platform='HM450')
```

### Differential methylation

test differential methylation on each locus
```{r}
betas <- SeSAMeGetExample('HM450.betas.76matchedTCGAchr20')
sample.info <- SeSAMeGetExample('HM450.sampleinfo.76matchedTCGAchr20')
cf <- DMR(betas, sample.info, ~patient+type, platform='HM450')
```

top differentially methylated probes on factor "type"
```{r}
cf1 <- cf$typeTumour
head(topLoci(cf1))
```

top segments
```R
head(topSegments(cf1))
```
      
visualize top DMR
```R
visualizeProbes(rownames(cf1[cf1[,'Seg.ID']==topSegments(cf1)$Seg.ID[1],]), betas, upstream=5000, dwstream=5000, platform='HM450',heat.height=3.5)
```


### CNV
    
uses LNCaP EPIC data from GSE86833
```{r}
example.sset <- SeSAMeGetExample('EPIC.sset.LNCaP.Rep1')
segs <- cnSegmentation(example.sset)
```

To visualize segments,
```{r}
visualizeSegments(segs)
```

### cell composition deconvolution

Use blood set as example,

```{r}
g <- diffRefSet(getRefSet(c('CD4T','CD19B','CD14Monocytes','CD56NK', 'granulocytes'), platform='HM450'))
betas <- SeSAMeGetExample('HM450.betas.TCGA-2L-AAQA-01A-21D-A38H-05')
estimateCellComposition(g, betas[rownames(g)])$frac
```
